### PR TITLE
feat: adding custom fee token support for orbit chains

### PR DESCRIPTION
**Pull request by GreatSoshiant**
In this PR we've added bunch of features to enable Orbit chains with custom fee tokens. Some of them here are:
1. erc20 token bridge deployment
2. Transfering native ERC20 token instead of ETH in ERC20 Path
3. adding a path to deposit more native token (similar to charging more ETH)

### CODE REVIEW

File: scripts/setup.ts [hunk 1]
```diff
@@ -119,7 +120,7 @@ async function main() {
         const newBalance = await L3Provider.getBalance(config.chainOwner)
```
Comment (Jason-W123) [inline scripts/setup.ts:120]: If it is erc20 deposit, here we should check erc20's balance?

File: toekn-bridge-contracts/AbsInboxMock.d.ts
[the remark this answers is not in the document]
Comment (gzeoneth) [reply toekn-bridge-contracts/AbsInboxMock.d.ts:1]: These type files should not be included, it should import from the packages like this
https://github.com/OffchainLabs/orbit-setup-script/blob/a6159b0532f4fa39cb21a9e42385ce66d06e21ad/scripts/tokenBridgeDeployment.ts#L19-L20
Will ping when the relevant package is released later today.

**Comment by gzeoneth** [comment]
Since https://github.com/OffchainLabs/token-bridge-contracts/releases/tag/v1.1.0-beta.1 is now released, we can use it instead of those type files.

File: scripts/createTokenBridge.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
+import { JsonRpcProvider } from '@ethersproject/providers'
+import { L1Network, L2Network, addCustomNetwork } from '@arbitrum/sdk'
+import { RollupAdminLogic__factory } from '@arbitrum/sdk/dist/lib/abi/factories/RollupAdminLogic__factory'
+import { createTokenBridge, getSigner } from './erc20TokenBridgeDeployment'
+import { L1AtomicTokenBridgeCreator__factory } from '../toekn-bridge-contracts'
+import * as fs from 'fs'
+
+const TOKEN_BRIDGE_CREATOR_Arb_Goerli =
```
Comment (gvladika) [inline scripts/createTokenBridge.ts:8]: We can probably switch now to the "official" release - `0x17412CC654a49Cdd5cE6965359d190F100Cf24d9` is token bridge creator on ArbGoerli, and `0x0025E5A25D64e02Fa711cfAEdf83987dac917eaC` on ArbSepolia

File: scripts/chargeEthOrErc20.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    // deposit 0.4 ETH
+    const tx = await contract.depositEth({
+      value: ethers.utils.parseEther('0.4'),
+    })
+    console.log('Transaction hash on parent chain: ', tx.hash)
+    await tx.wait()
+    console.log('Transaction has been mined')
+    console.log('0.4 ETHs are deposited to your account')
+  } else {
+    tx = await erc20Inbox.depositERC20(ethers.utils.parseEther(amount))
```
Comment (gvladika) [inline scripts/chargeEthOrErc20.ts:68]: This will require native tokens to be approved earlier

File: scripts/erc20TokenBridgeDeployment.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    await l1TokenBridgeCreator.getCanonicalL2CustomGatewayAddress(childChainId)
+
+  const isUsingFeeToken = feeToken != ethers.constants.AddressZero
+  const l2WethGateway = isUsingFeeToken
+    ? ethers.constants.AddressZero
+    : L2WethGateway__factory.attach(
+        await l1TokenBridgeCreator.getCanonicalL2WethGatewayAddress(
+          childChainId
+        )
+      ).connect(l2Provider).address
```
Comment (gzeoneth) [inline scripts/erc20TokenBridgeDeployment.ts:212]:
```suggestion
    : await l1TokenBridgeCreator.getCanonicalL2WethGatewayAddress(
          childChainId
        )
```

File: scripts/erc20TokenBridgeDeployment.ts [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+  L2AtomicTokenBridgeFactory
+)
+
+import L2GatewayRouter from '@arbitrum/token-bridge-contracts/build/contracts/contracts/tokenbridge/arbitrum/gateway/L2GatewayRouter.sol/L2GatewayRouter.json'
+import L2ERC20Gateway from '@arbitrum/token-bridge-contracts/build/contracts/contracts/tokenbridge/arbitrum/gateway/L2ERC20Gateway.sol/L2ERC20Gateway.json'
+const L2ERC20Gateway__factory = NamedFactoryInstance(L2ERC20Gateway)
+import L2CustomGateway from '@arbitrum/token-bridge-contracts/build/contracts/contracts/tokenbridge/arbitrum/gateway/L2CustomGateway.sol/L2CustomGateway.json'
+
+import L2WethGateway from '@arbitrum/token-bridge-contracts/build/contracts/contracts/tokenbridge/arbitrum/gateway/L2WethGateway.sol/L2WethGateway.json'
+const L2WethGateway__factory = NamedFactoryInstance(L2WethGateway)
```
Comment (gzeoneth) [inline scripts/erc20TokenBridgeDeployment.ts:42]:
```suggestion
```

**Review by gzeoneth — COMMENTED** [review]
Awesome, it looks much cleaner now. Some minor comment and some test to make sure both the ETH and ERC20 rollup setup works and we are good to go!

**Review by gzeoneth — APPROVED** [review]
LGTM, thanks @Jason-W123 for testing

**Review by gvladika — APPROVED** [review]

**Review by gzeoneth — CHANGES_REQUESTED** [review]
the token bridge creator lacked a getter for canonicalL2Multicall address
we also need a L1Multicall (not ArbMulticall) to be deployed on the parent chain
^^^ blockchain eng is working on fixing these

the setup script also lacked outputInfo.json which was previously generated by tokenBridgeDeployment.ts

File: scripts/erc20TokenBridgeDeployment.ts [hunk 6]
```diff
@@ -219,6 +219,11 @@ export const createTokenBridge = async (
   const l2ProxyAdmin =
     await l1TokenBridgeCreator.getCanonicalL2ProxyAdminAddress(childChainId)
 
+  const l1MultiCall = await l1TokenBridgeCreator.l1MultiCall()
```
Comment (gvladika) [inline scripts/erc20TokenBridgeDeployment.ts:222]: Typo in function name
```suggestion
  const l1MultiCall = await l1TokenBridgeCreator.l1Multicall()
```

**Comment by gzeoneth** [comment]
Pre-release checklist:

- [ ] Merge https://github.com/OffchainLabs/orbit-setup-script/pull/30
- [ ] Test rollup deployment flow
- [ ] Test AnyTrust deployment flow
- [ ] Test custom fee token deployment flow
- [ ] Verify validator works in all deployment types
- [ ] Verify token bridge works in all deployment types
- [ ] Validate rollup ownership after setup
- [ ] Validate tokenbridge ownership after setup

**Review by gzeoneth — APPROVED** [review]

**Pull request merged**